sentry/src/main/java/io/sentry/MetricsAggregator.java

```diff
@@ -0,0 +1,320 @@
+      add(MetricType.Distribution, key, value, unit, tags, startMs, stackLevel + 1);
+    }
+  }
+
+  @SuppressWarnings({"FutureReturnValueIgnored", "UnusedVariable"})
+  private void add(
+      final @NotNull MetricType type,
+      final @NotNull String key,
+      final double value,
+      @Nullable MeasurementUnit unit,
+      final @Nullable Map<String, String> tags,
+      @Nullable Long timestampMs,
+      final int stackLevel) {
+
+    if (isClosed) {
+      return;
+    }
+
+    if (timestampMs == null) {
+      timestampMs = timeProvider.getTimeMillis();
+    }
+
+    final @NotNull Map<String, String> enrichedTags = enrichTags(tags);
+
+    final @NotNull Metric metric;
+    switch (type) {
+      case Counter:
+        metric = new CounterMetric(key, value, unit, enrichedTags, timestampMs);
+        break;
+      case Gauge:
+        metric = new GaugeMetric(key, value, unit, enrichedTags, timestampMs);
+        break;
+      case Distribution:
+        metric = new DistributionMetric(key, value, unit, enrichedTags, timestampMs);
+        break;
+      case Set:
+        metric = new SetMetric(key, unit, enrichedTags, timestampMs);
+        //noinspection unchecked
+        metric.add((int) value);
+        break;
+      default:
+        throw new IllegalArgumentException("Unknown MetricType: " + type.name());
+    }
+
+    final long timeBucketKey = MetricsHelper.getTimeBucketKey(timestampMs);
+    final @NotNull Map<String, Metric> timeBucket = getOrAddTimeBucket(timeBucketKey);
+
+    final @NotNull String metricKey =
+        MetricsHelper.getMetricBucketKey(type, key, unit, enrichedTags);
+
+    // TODO check if we can synchronize only the metric itself
+    synchronized (timeBucket) {
+      @Nullable Metric existingMetric = timeBucket.get(metricKey);
+      if (existingMetric != null) {
+        existingMetric.add(value);
+      } else {
+        timeBucket.put(metricKey, metric);
+      }
+    }
+
+    // spin up real executor service the first time metrics are collected
+    if (!isClosed && !flushScheduled) {
+      synchronized (this) {
+        if (!isClosed && !flushScheduled) {
+          flushScheduled = true;
+          // TODO this is probably not a good idea after all
+          // as it will slow down the first metric emission
+          // maybe move to constructor?
+          if (executorService instanceof NoOpSentryExecutorService) {
+            executorService = new SentryExecutorService();
+          }
+          executorService.schedule(this, MetricsHelper.FLUSHER_SLEEP_TIME_MS);
+        }
+      }
+    }
+  }
+
+  @NotNull
+  private Map<String, String> enrichTags(final @Nullable Map<String, String> tags) {
+    final @NotNull Map<String, String> defaultTags = hub.getDefaultTagsForMetric();
+    if (tags == null) {
+      return Collections.unmodifiableMap(defaultTags);
+    }
+
+    final @NotNull Map<String, String> enrichedTags = new HashMap<>(tags);
```

Comment:
can we use `tags` directly instead of creating a new map?

Reply:
As the `tags` are provided by the developer, they could be immutable, thus I think it's safer to clone them